server/admin/server_test.go

```diff
@@ -76,46 +80,86 @@ func genCertPair(certFile, keyFile string) error {
 
 var tPort = 5555
 
-func newTServer(t *testing.T, start bool, authSHA [32]byte) (*Server, *TCore, func()) {
-	c := &TCore{}
-	ctx, cancel := context.WithCancel(context.Background())
-	tmp, err := os.Getwd()
+// If start is true, the Server's Run goroutine is started, and the shutdown
+// func must be called when finished with the Server.
+func newTServer(t *testing.T, start bool, authSHA [32]byte) (*Server, func()) {
+	tmp, err := ioutil.TempDir("", "admin")
 	if err != nil {
 		t.Fatal(err)
 	}
-	cert, key := tmp+"/cert.cert", tmp+"/key.key"
+	defer os.RemoveAll(tmp)
+
+	cert, key := filepath.Join(tmp, "tls.cert"), filepath.Join(tmp, "tls.key")
 	err = genCertPair(cert, key)
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer os.Remove(cert)
-	defer os.Remove(key)
-	cfg := &SrvConfig{
-		Core:    c,
+
+	s, err := NewServer(&SrvConfig{
+		Core:    new(TCore),
 		Addr:    fmt.Sprintf("localhost:%d", tPort),
 		Cert:    cert,
 		Key:     key,
 		AuthSHA: authSHA,
-	}
-	s, err := NewSrv(cfg)
+	})
 	if err != nil {
-		t.Fatalf("error creating server: %v", err)
+		t.Fatalf("error creating Server: %v", err)
 	}
-	if start {
+	if !start {
+		return s, func() {}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
 		s.Run(ctx)
+		wg.Done()
+	}()
+	shutdown := func() {
+		cancel()
+		wg.Wait()
+	}
+	return s, shutdown
```

Comment:
Otherwise calling `cancel` didn't wait for shutdown, it just proceeded to the next test or returned to os.